### PR TITLE
Playback at "fast" speed

### PR DESCRIPTION
**Pull request by jbrains**
By pressing "G", the viewer can play the video at a "fast" speed. If people like this, then we could give them a configuration setting to choose their own "fast" speed.

I wanted this because I was watching a bunch of online training videos and I wanted to just jump to 2.5x speed for every new video in the course.

### CODE REVIEW

**Comment by xftroxgpx** [comment]
view without whitespace: https://github.com/igrigorik/videospeed/pull/203/files?w=1

**Comment by jbrains** [comment]
@xftroxgpx Yes. I chose to err on the side of keeping those comments lined up, since I didn't know how much you folks cared about it. :)

**Comment by igrigorik** [comment]
> I wanted this because I was watching a bunch of online training videos and I wanted to just jump to 2.5x speed for every new video in the course.

Curious, why does "remember speed" setting fall short for this use case? If you have that enabled it should store the last playback speed and automatically start the next one at same speed.

**Comment by jbrains** [comment]
@igrigorik I suppose I think of it like a "preset speed", like a favorite station on the radio. I suppose I could use "remember speed" for this purpose, but if I want to go faster or slower on this one video, it's nice to quickly jump back to "preferred fast speed" instead of having to adjust.

**Comment by igrigorik** [comment]
Fair enough, this is a feature that's been requested a few times in the past..

As an aside, I'm wary of adding more and more custom shortcuts -- this process doesn't scale. One escape hatch I can see here is that we can provide a "create your own shortcut" for power users such as yourself: expose an interface to define a key and a command that should be run. This way we can get out of business of writing more and more one-offs and instead let the power users do whatever they need it to do. That said, in this particular case, I thin we can make an exception. I have not personally felt the need for this, but I've heard the request enough times from others..

File: inject.js [hunk 1]
```diff
@@ -1,19 +1,21 @@
...
+      fastSpeed: 2.5,                // default 2.5x
+      rewindTime: 10,                // default 10s
+      advanceTime: 10,               // default 10s
+      resetKeyCode:  82,             // default: R
+      slowerKeyCode: 83,             // default: S
+      fasterKeyCode: 68,             // default: D
+      rewindKeyCode: 90,             // default: Z
+      advanceKeyCode: 88,            // default: X
+      displayKeyCode: 86,            // default: V
+      playAtFastSpeedKeyCode: 71,    // default: G
```
Comment (igrigorik) [inline inject.js]: Let's do `fastKeyCode` to be consistent with the rest.

File: inject.js [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
-      fasterKeyCode: 68,    // default: D
-      rewindKeyCode: 90,    // default: Z
-      advanceKeyCode: 88,   // default: X
-      displayKeyCode: 86,   // default: V
-      rememberSpeed: false, // default: false
-      startHidden: false,   // default: false
+      speed: 1.0,                    // default 1x
+      resetSpeed: 1.0,               // default 1x
+      speedStep: 0.1,                // default 0.1x
+      fastSpeed: 2.5,                // default 2.5x
```
Comment (igrigorik) [inline inject.js]: As a default, I think most folks will find this too fast. I recommend we set 1.75 as the default, which is the recommended "comfortable" speed for most folks based on past research.

**Comment by jbrains** [comment]
@igrigorik I understand. I think one could extract it quite naturally from what's already there; I imagine it's just waiting for someone really eager to do it. :)

Thanks.

File: inject.js [hunk 3]
```diff
@@ -321,11 +325,18 @@ chrome.extension.sendMessage({}, function(response) {
...
         } else if (action === 'drag') {
           handleDrag(v, controller);
+        } else if (action === 'fast') {
+          playVideoAtFastSpeed(v);
         }
       }
     });
   }
 
+  // CONTRACT: video has a way to set the playback rate.
```
Comment (igrigorik) [inline inject.js]: What do you mean by this? All `video` elements can do this..

Comment (jbrains) [inline inject.js]: Just documenting the part of `video` that that function cares about. My convention. I'm happy to remove it.

File: inject.js (hunk 2 above)
Comment (jbrains) [inline inject.js]: When I add the configuration ability, I'll change this to 1.75.

File: inject.js (hunk 3 above)
Comment (igrigorik) [inline inject.js]: Gotcha. Yeah, let's drop it, it's implicit here and stands out as a red flag with respect to the rest of the file.

**Comment by jbrains** [comment]
I think I've made the changes you'd asked for, so it's ready for your next review.

**Comment by xftroxgpx** [comment]
One thing: if you set it to 1.75, then you can't get back to 1.00 (since the decrements are by 0.10). Maybe someone(me?) should make a new PR to ensure that the next inc/dec is to a rounded value (but still no more than 0.10 per step)? then 1.75 would be great here(next would be 1.80 then 1.90 or prev. would be 1.70 then 1.60), and it would also handle this case https://github.com/igrigorik/videospeed/pull/200

File: inject.js [hunk 4]
```diff
@@ -1,19 +1,21 @@
...
-      advanceTime: 10,      // default 10s
-      resetKeyCode:  82,    // default: R
-      slowerKeyCode: 83,    // default: S
-      fasterKeyCode: 68,    // default: D
-      rewindKeyCode: 90,    // default: Z
-      advanceKeyCode: 88,   // default: X
-      displayKeyCode: 86,   // default: V
-      rememberSpeed: false, // default: false
-      startHidden: false,   // default: false
+      speed: 1.0,                    // default 1x
```
Comment (igrigorik) [inline inject.js]: Nitpick: can you please reverse the whitespace change to original?

File: options.html [hunk 5]
```diff
@@ -33,6 +33,10 @@
         <input id="fasterKeyInput" placeholder="press a key" type="text" value=""/>
       </div>
       <div class="row">
+        <label for="fastKeyInput">Jump to Preferred Fast speed</label>
```
Comment (igrigorik) [inline options.html]: Let's simplify this to "Preferred speed"

File: options.html [hunk 6]
```diff
@@ -57,6 +61,10 @@
           <input id="speedStep" type="text" value=""/>
       </div>
       <div class="row">
+          <label for="fastSpeed">Preferred Fast Playback Speed (x)</label>
```
Comment (igrigorik) [inline options.html]: Ditto, "Preferred Speed". 

(and I just realized we flip casing between shortcuts and settings.. I'll fix this later)

**Comment by igrigorik** [comment]
@xftroxgpx @jbrains I'm good with 1.8 as default.

File: inject.js (hunk 4 above)
Comment (jbrains) [inline inject.js]: Easily enough.

**Comment by jbrains** [comment]
I think I got everything. It's possible that @igrigorik meant something slightly different with the configuration settings labels, but otherwise I do believe I got it all.

**Comment by igrigorik** [comment]
Looks great, thanks for your patience! Merging.. :-)

**Pull request merged**